File: Python/smallest-subarray-to-sort-in-every-sliding-window.py
# Time:  O(n)
# Space: O(n)

# mono stack, two pointers
class Solution(object):
    def minSubarraySort(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        def count(nums):
            nxt, stk = [n]*n, []
            for i in reversed(xrange(n)):
                while not (not stk or nums[stk[-1]] >= nums[i]):
                    stk.pop()
                if stk:
                    nxt[i] = stk[-1]
                stk.append(i)
            result = []
            j, left = 0, -1
            for i in xrange(1, n):
                if nums[i] < nums[i-1]:
                    left = i
                if i < k-1:
                    continue
                j = max(j, i-(k-1))
                while not nxt[j] > left:
                    j = nxt[j]  # or j += 1
                result.append(max(i-nxt[j]+1, 0))
            return result

        n = len(nums)
        if k == 1:
            return [0]*(n-k+1)
        right = count(nums)
        for i in xrange((n+1)//2):
            nums[i], nums[~i] = -nums[~i], -nums[i]
        left = count(nums)
        return [max(k-left[~i]-right[i], 0) for i in xrange(n-k+1)]

File: C++/smallest-subarray-to-sort-in-every-sliding-window.cpp
// Time:  O(n)
// Space: O(n)

// mono stack, two pointers
class Solution {
public:
    vector<int> minSubarraySort(vector<int>& nums, int k) {
        const int n = size(nums);
        const auto& count = [&](const auto& nums) {
            vector<int> nxt(n, n), stk;
            for (int i = n - 1; i >= 0; --i) {
                while (!(empty(stk) || nums[stk.back()] >= nums[i])) {
                    stk.pop_back();
                }
                if (!empty(stk)) {
                    nxt[i] = stk.back();
                }
                stk.emplace_back(i);
            }
            vector<int> result;
            for (int i = 1, j = 0, left = -1; i < n; ++i) {
                if (nums[i] < nums[i - 1]) {
                    left = i;
                }
                if (i < k - 1) {
                    continue;
                }
                j = max(j, i - (k - 1));
                while (!(nxt[j] > left)) {
                    j = nxt[j];  // or ++j
                }
                result.emplace_back(max(i - nxt[j] + 1, 0));
            }
            return result;
        };

        vector<int> result(n - k + 1);
        if (k == 1) {
            return result;
        }
        const auto& right = count(nums);
        for (int i = 0; i <= n - 1 - i; ++i) {
            tie(nums[i], nums[n - 1 - i]) = pair(-nums[n - 1 - i], -nums[i]);
        }
        const auto& left = count(nums);
        for (int i = 0; i < size(result); ++i) {
            result[i] = max(k - left[((n - k + 1) - 1) - i] - right[i], 0);
        }
        return result;
    }
};
